src/lanczos/lanczos.jl

```diff
@@ -0,0 +1,105 @@
+using Base.Cartesian
+using StaticArrays
+
+export Lanczos
+
+"""
+    Lanczos(a=4, n=a)
+
+Lanczos resampling via a kernel with scale parameter `a` and support over `n` neighbors.
+
+This form of interpolation is merely the discrete convolution of the samples with a Lanczos kernel of size `a`. The size is directly related to how "far" the interpolation will reach for information, and has `O(n^2)` impact on runtime. A default value of 4 matches the OpenCV implementation `lanczos4`.
+"""
+struct Lanczos <: InterpolationType
+    a::Int
+    n::Int
+    
+    function Lanczos(a, n)
+        n < a && @warn "Using a smaller support than scale for Lanczos window. Proceed with caution."
+        new(a, n)
+    end
+end
+
+Lanczos(a=4) = Lanczos(a, a)
+
+"""
+    LanczosInterpolation
+"""
+struct LanczosInterpolation{T,N,IT <: DimSpec{Lanczos},A <: AbstractArray{T,N},P <: Tuple{Vararg{AbstractArray,N}}} <: AbstractInterpolation{T,N,IT}
+    coefs::A
+    parentaxes::P
+    it::IT
+end
+
+getknots(itp::LanczosInterpolation) = axes(itp)
+coefficients(itp::LanczosInterpolation) = itp.coefs
+itpflag(itp::LanczosInterpolation) = itp.it
+
+size(itp::LanczosInterpolation) = map(length, itp.parentaxes)
+axes(itp::LanczosInterpolation) = itp.parentaxes
+lbounds(itp::LanczosInterpolation) = map(first, itp.parentaxes)
+ubounds(itp::LanczosInterpolation) = map(last, itp.parentaxes)
+
+function interpolate(A::AbstractArray{T}, it::Lanczos) where T
+    Apad = copy_with_padding(float(T), A, it)
+    return LanczosInterpolation(Apad, axes(A), it)
+end
+
+@inline function (itp::LanczosInterpolation{T,N})(x::Vararg{<:Number,N}) where {T,N}
+    @boundscheck (checkbounds(Bool, itp, x...) || Base.throw_boundserror(itp, x))
+    wis = weightedindexes((value_weights,), itpinfo(itp)..., x)
+    itp.coefs[wis...]
+end
+
+function weightedindex_parts(fs, it::Lanczos, ax::AbstractUnitRange{<:Integer}, x)
+    pos, δx = positions(it, ax, x)
+    (position = pos, coefs = fmap(fs, it, δx))
+end
+
+function positions(it::Lanczos, ax, x)
+    xf = floorbounds(x, ax)
+    δx = x - xf
+    fast_trunc(Int, xf) - it.n + 1, δx
+end
+
+function value_weights(it::Lanczos, δx::S) where S
+    T = float(S)
+    # short-circuit if integral
+    isinteger(δx) && return ntuple(i -> i == it.n - δx ? one(T) : zero(T), 2it.n)
```

Comment:
@mkitti, not sure where you are currently on your Julia knowledge. Let's proceed by hints to give you a chance to figure out as much as you can on your own. This first hint is that the performance issue mentioned in the discussion is on this line. First task is to describe what about this line contributes to poor performance. Second task is to come up with an understanding of what might be done about it, along with the likely tradeoffs of that solution. (Do your best! I'm here to help, but you'll get there faster if you put effort into solving this yourself.)

Reply:
My understanding is there are several specializations issues here.

1. One issue is that given type `Lanczos` and `S` the size of the ntuple that is variable and cannot be predicted. My suggestion above is to parameterize `Lanczos` on n. That is given input types to value_weights of  `Lanczos{N}` and `S` the output of this function is a tuple of length 2N. The risk is over-specialization. There would be a distinct version of `value_weights` for each N. I think this is the main issue.

2. Another issue is there could be some optimization of the anonymous function and use of ntuple. `it.n -δx` is a constant and indicates the exact index where the tuple should be 1. `dx` could be converted to an integer. The last argument could be wrapped by `Val` to specialize `ntuple` and stabilize the return type.

Reply:
Very good. Your first point hits all the main points, both in terms of the solution and its downsides. @mileslucas might be able to tell us if the diversity of lengths would be a major problem.

For the second one, wrapping an ordinary number in `Val` doesn't help, you need to be able to predict it from the beginning. See https://docs.julialang.org/en/latest/manual/performance-tips/#man-performance-value-type-1.

Reply:
Regarding `Val`, I think that depends on `N`:

https://github.com/JuliaLang/julia/blob/master/base/ntuple.jl#L17-L73

```
function ntuple(f::F, n::Integer) where F
    t = n == 0  ? () :
        n == 1  ? (f(1),) :
        n == 2  ? (f(1), f(2)) :
        n == 3  ? (f(1), f(2), f(3)) :
        n == 4  ? (f(1), f(2), f(3), f(4)) :
        n == 5  ? (f(1), f(2), f(3), f(4), f(5)) :
        n == 6  ? (f(1), f(2), f(3), f(4), f(5), f(6)) :
        n == 7  ? (f(1), f(2), f(3), f(4), f(5), f(6), f(7)) :
        n == 8  ? (f(1), f(2), f(3), f(4), f(5), f(6), f(7), f(8)) :
        n == 9  ? (f(1), f(2), f(3), f(4), f(5), f(6), f(7), f(8), f(9)) :
        n == 10 ? (f(1), f(2), f(3), f(4), f(5), f(6), f(7), f(8), f(9), f(10)) :
        _ntuple(f, n)
    return t
end

function _ntuple(f, n)
    @_noinline_meta
    (n >= 0) || throw(ArgumentError(string("tuple length should be ≥ 0, got ", n)))
    ([f(i) for i = 1:n]...,)
end

# inferrable ntuple (enough for bootstrapping)
ntuple(f, ::Val{0}) = ()
ntuple(f, ::Val{1}) = (@_inline_meta; (f(1),))
ntuple(f, ::Val{2}) = (@_inline_meta; (f(1), f(2)))
ntuple(f, ::Val{3}) = (@_inline_meta; (f(1), f(2), f(3)))


"""
    ntuple(f, ::Val{N})
Create a tuple of length `N`, computing each element as `f(i)`,
where `i` is the index of the element. By taking a `Val(N)`
argument, it is possible that this version of ntuple may
generate more efficient code than the version taking the
length as an integer. But `ntuple(f, N)` is preferable to
`ntuple(f, Val(N))` in cases where `N` cannot be determined
at compile time.
# Examples
```jldoctest
julia> ntuple(i -> 2*i, Val(4))
(2, 4, 6, 8)
"""
@inline function ntuple(f::F, ::Val{N}) where {F,N}
    N::Int
    (N >= 0) || throw(ArgumentError(string("tuple length should be ≥ 0, got ", N)))
    if @generated
        quote
            @nexprs $N i -> t_i = f(i)
            @ncall $N tuple t
        end
    else
        Tuple(f(i) for i = 1:N)
    end
end
```

I think @mileslucas is mainly concerned up to Lanczos{4} up to , which would be mean N = 8 for `ntuple`

Reply:
julia> using BenchmarkTools

julia> test_ntuple(N,x) = ntuple(i -> i == N - x, 2N)
test_ntuple (generic function with 2 methods)

julia> test_ntuple_pass_N(v::Val{N},x) where N = ntuple(i -> i == N/2 - x, N)
test_ntuple_pass_N (generic function with 1 method)

julia> test_ntuple_val(v::Val{N},x) where N = ntuple(i -> i == N/2 - x, v)
test_ntuple_val (generic function with 2 methods)

julia> let N = Ref(2), x = Ref(1.0), v = Ref(Val(N[]*2))
           @btime $test_ntuple($N[], $x[])
           @btime $test_ntuple_pass_N($v[], $x[])
           @btime $test_ntuple_val($v[], $x[])
       end
  19.670 ns (1 allocation: 16 bytes)
  19.102 ns (1 allocation: 16 bytes)
  5.727 ns (0 allocations: 0 bytes)
(true, false, false, false)

julia> let N = Ref(3), x = Ref(1.0), v = Ref(Val(N[]*2))
           @btime $test_ntuple($N[], $x[])
           @btime $test_ntuple_pass_N($v[], $x[])
           @btime $test_ntuple_val($v[], $x[])
       end
  21.712 ns (1 allocation: 16 bytes)
  21.182 ns (1 allocation: 16 bytes)
  8.882 ns (0 allocations: 0 bytes)
(false, true, false, false, false, false)


julia> let N = Ref(4), x = Ref(1.0), v = Ref(Val(N[]*2))
           @btime $test_ntuple($N[], $x[])
           @btime $test_ntuple_pass_N($v[], $x[])
           @btime $test_ntuple_val($v[], $x[])
       end
  23.937 ns (1 allocation: 16 bytes)
  23.129 ns (1 allocation: 16 bytes)
  11.213 ns (0 allocations: 0 bytes)
(false, false, true, false, false, false, false, false)

Reply:
It's not just the construction of the tuple; it's also the *consumer* of the tuple. Consider:
```julia
julia> struct HideInt
           n::Int
       end

julia> buildt(::Val{N}) where N = ntuple(identity, Val(N))
buildt (generic function with 1 method)

julia> buildt(h::HideInt) = ntuple(identity, h.n)
buildt (generic function with 2 methods)

julia> f(x) = sum(buildt(x))
f (generic function with 1 method)

julia> using BenchmarkTools
[ Info: Precompiling BenchmarkTools [6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf]

julia> @btime f(x) setup=(x=Val(6))    # this result is "fake"
  0.016 ns (0 allocations: 0 bytes)
21

julia> @btime f(x) setup=(x=Val(rand(1:6)))
  7.640 ns (0 allocations: 0 bytes)
1

julia> @btime f(x) setup=(x=HideInt(rand(1:6)))
  14.472 ns (0 allocations: 0 bytes)
```

and

```julia
julia> @code_warntype f(Val(5))
Variables
  #self#::Core.Compiler.Const(f, false)
  x::Core.Compiler.Const(Val{5}(), false)

Body::Int64
1 ─ %1 = Main.buildt(x)::Core.Compiler.Const((1, 2, 3, 4, 5), false)
│   %2 = Main.sum(%1)::Core.Compiler.Const(15, false)
└──      return %2

julia> @code_warntype f(HideInt(5))
Variables
  #self#::Core.Compiler.Const(f, false)
  x::HideInt

Body::Any
1 ─ %1 = Main.buildt(x)::Tuple{Vararg{Int64,N} where N}
│   %2 = Main.sum(%1)::Any
└──      return %2
```

Reply:
Also, use backticks to avoid pinging any user named `@Btime`

Reply:
This is getting a bit long, but parameterizing `Lanczos` alone does not appear to create a specific return type. I just get `Tuple`:

```
julia> import Interpolations: InterpolationType, _lanczos4, lanczos

julia> struct Lanczos{N} <: InterpolationType where {N <: Int}
           a::Int

           function Lanczos{N}(a) where {N}
               N < a && @warn "Using a smaller support than scale for Lanczos window. Proceed with caution."
               new(a)
           end
       end

julia> Lanczos(a=4) = Lanczos{a}(a)
Lanczos

julia> function value_weights(it::Lanczos{N}, δx::S) where {N,S}
           T = float(S)
           # short-circuit if integral
           isinteger(δx) && return ntuple(i -> i == N - δx ? one(T) : zero(T), 2N)
           # LUTs
           it.a == N == 4 && return _lanczos4(δx)

           cs = ntuple(i -> lanczos(N - i + δx, it.a, N), 2N)
           normed_cs = ntuple(i -> cs[i] / sum(cs), length(cs))
           return normed_cs
       end
value_weights (generic function with 1 method)

julia> `@code_warntype` value_weights(Lanczos(2),1.0)
Variables
  #self#::Core.Compiler.Const(value_weights, false)
  it::Lanczos{2}
  δx::Float64
  #25::var"#25#28"{2,Float64,DataType}
  #26::var"#26#29"{2,Lanczos{2},Float64}
  #27::var"#27#30"{_A} where _A
  T::Type{Float64}
  cs::Tuple{Vararg{Float64,N} where N}
  normed_cs::Tuple
  @_10::Bool

Body::Tuple
1 ─       Core.NewvarNode(:(#25))
│         Core.NewvarNode(:(#26))
│         Core.NewvarNode(:(#27))
│         Core.NewvarNode(:(cs))
│         Core.NewvarNode(:(normed_cs))
│         (T = Main.float($(Expr(:static_parameter, 2))))
│   %7  = Main.isinteger(δx)::Bool
└──       goto #3 if not %7
2 ─ %9  = Main.:(var"#25#28")::Core.Compiler.Const(var"#25#28", false)
│   %10 = $(Expr(:static_parameter, 1))::Core.Compiler.Const(2, false)
│   %11 = Core.typeof(δx)::Core.Compiler.Const(Float64, false)
│   %12 = Core.typeof(T::Core.Compiler.Const(Float64, false))::Core.Compiler.Const(DataType, false)
│   %13 = Core.apply_type(%9, %10, %11, %12)::Core.Compiler.Const(var"#25#28"{2,Float64,DataType}, false)
│         (#25 = %new(%13, δx, T::Core.Compiler.Const(Float64, false)))
│   %15 = #25::var"#25#28"{2,Float64,DataType}
│   %16 = (2 * $(Expr(:static_parameter, 1)))::Core.Compiler.Const(4, false)
│   %17 = Main.ntuple(%15, %16)::Tuple
└──       return %17
3 ─ %19 = Base.getproperty(it, :a)::Int64
│   %20 = (%19 == $(Expr(:static_parameter, 1)))::Bool
└──       goto #5 if not %20
4 ─       (@_10 = $(Expr(:static_parameter, 1)) == 4)
└──       goto #6
5 ─       (@_10 = false)
6 ┄       goto #8 if not @_10::Core.Compiler.Const(false, false)
7 ─       Core.Compiler.Const(:(Main._lanczos4(δx)), false)
└──       Core.Compiler.Const(:(return %26), false)
8 ┄ %28 = Main.:(var"#26#29")::Core.Compiler.Const(var"#26#29", false)
│   %29 = $(Expr(:static_parameter, 1))::Core.Compiler.Const(2, false)
│   %30 = Core.typeof(it)::Core.Compiler.Const(Lanczos{2}, false)
│   %31 = Core.typeof(δx)::Core.Compiler.Const(Float64, false)
│   %32 = Core.apply_type(%28, %29, %30, %31)::Core.Compiler.Const(var"#26#29"{2,Lanczos{2},Float64}, false)
│         (#26 = %new(%32, it, δx))
│   %34 = #26::var"#26#29"{2,Lanczos{2},Float64}
│   %35 = (2 * $(Expr(:static_parameter, 1)))::Core.Compiler.Const(4, false)
│         (cs = Main.ntuple(%34, %35))
│   %37 = Main.:(var"#27#30")::Core.Compiler.Const(var"#27#30", false)
│   %38 = Core.typeof(cs)::Type{Tuple{Vararg{Float64,N}}} where N
│   %39 = Core.apply_type(%37, %38)::Type{var"#27#30"{_A}} where _A
│         (#27 = %new(%39, cs))
│   %41 = #27::var"#27#30"{_A} where _A
│   %42 = Main.length(cs)::Int64
│         (normed_cs = Main.ntuple(%41, %42))
└──       return normed_cs
```

Reply:
Adding in `Val` seems more stable. I get `NTuple{4,Float64}`:

```
julia> function value_weights(it::Lanczos{N}, δx::S) where {S,N}
           # short-circuit if integral
           isinteger(δx) && return ntuple(i->convert(float(S),i == N - δx), Val(2N))

           # LUTs
           it.a == N == 4 && return _lanczos4(δx)

           cs = ntuple(i -> lanczos(N - i + δx, it.a, N), Val(2N))
           normed_cs = ntuple(i -> cs[i] / sum(cs), Val(length(cs)))
           return normed_cs
       end
value_weights (generic function with 1 method)

julia> @code_warntype value_weights(Lanczos(2),1.0)
Variables
  #self#::Core.Compiler.Const(value_weights, false)
  it::Lanczos{2}
  δx::Float64
  #43::var"#43#46"{2,Float64,Float64}
  #44::var"#44#47"{2,Lanczos{2},Float64}
  #45::var"#45#48"{NTuple{4,Float64}}
  cs::NTuple{4,Float64}
  normed_cs::NTuple{4,Float64}
  @_9::Bool

Body::NTuple{4,Float64}
1 ─       Core.NewvarNode(:(#43))
│         Core.NewvarNode(:(#44))
│         Core.NewvarNode(:(#45))
│         Core.NewvarNode(:(cs))
│         Core.NewvarNode(:(normed_cs))
│   %6  = Main.isinteger(δx)::Bool
└──       goto #3 if not %6
2 ─ %8  = Main.:(var"#43#46")::Core.Compiler.Const(var"#43#46", false)
│   %9  = $(Expr(:static_parameter, 2))::Core.Compiler.Const(2, false)
│   %10 = $(Expr(:static_parameter, 1))::Core.Compiler.Const(Float64, false)
│   %11 = Core.typeof(δx)::Core.Compiler.Const(Float64, false)
│   %12 = Core.apply_type(%8, %9, %10, %11)::Core.Compiler.Const(var"#43#46"{2,Float64,Float64}, false)
│         (#43 = %new(%12, δx))
│   %14 = #43::var"#43#46"{2,Float64,Float64}
│   %15 = (2 * $(Expr(:static_parameter, 2)))::Core.Compiler.Const(4, false)
│   %16 = Main.Val(%15)::Core.Compiler.Const(Val{4}(), true)
│   %17 = Main.ntuple(%14, %16)::NTuple{4,Float64}
└──       return %17
3 ─ %19 = Base.getproperty(it, :a)::Int64
│   %20 = (%19 == $(Expr(:static_parameter, 2)))::Bool
└──       goto #5 if not %20
4 ─       (@_9 = $(Expr(:static_parameter, 2)) == 4)
└──       goto #6
5 ─       (@_9 = false)
6 ┄       goto #8 if not @_9::Core.Compiler.Const(false, false)
7 ─       Core.Compiler.Const(:(Main._lanczos4(δx)), false)
└──       Core.Compiler.Const(:(return %26), false)
8 ┄ %28 = Main.:(var"#44#47")::Core.Compiler.Const(var"#44#47", false)
│   %29 = $(Expr(:static_parameter, 2))::Core.Compiler.Const(2, false)
│   %30 = Core.typeof(it)::Core.Compiler.Const(Lanczos{2}, false)
│   %31 = Core.typeof(δx)::Core.Compiler.Const(Float64, false)
│   %32 = Core.apply_type(%28, %29, %30, %31)::Core.Compiler.Const(var"#44#47"{2,Lanczos{2},Float64}, false)
│         (#44 = %new(%32, it, δx))
│   %34 = #44::var"#44#47"{2,Lanczos{2},Float64}
│   %35 = (2 * $(Expr(:static_parameter, 2)))::Core.Compiler.Const(4, false)
│   %36 = Main.Val(%35)::Core.Compiler.Const(Val{4}(), true)
│         (cs = Main.ntuple(%34, %36))
│   %38 = Main.:(var"#45#48")::Core.Compiler.Const(var"#45#48", false)
│   %39 = Core.typeof(cs)::Core.Compiler.Const(NTuple{4,Float64}, false)
│   %40 = Core.apply_type(%38, %39)::Core.Compiler.Const(var"#45#48"{NTuple{4,Float64}}, false)
│         (#45 = %new(%40, cs))
│   %42 = #45::var"#45#48"{NTuple{4,Float64}}
│   %43 = Main.length(cs)::Core.Compiler.Const(4, false)
│   %44 = Main.Val(%43)::Core.Compiler.Const(Val{4}(), true)
│         (normed_cs = Main.ntuple(%42, %44))
└──       return normed_cs
```

Reply:
Sorry for my delay here. In your first post, `where {N <: Int}` doesn't do what you are intending (really it should be `N::Int`, definitely not the same thing as `N <: Int`, but that's not supported).